automata/fa/animation.py

```diff
@@ -0,0 +1,380 @@
+    input_symbols : _ManimInput
+        The element in animation to show the current symbol of the `input_str`.
+    """
+
+    nfa: NFA
+    input_str: str
+    nfa_graph: _FAGraph
+    input_symbols: _ManimInput
+
+    def __init__(self, nfa: NFA, input_str: str, **kwargs: Any) -> None:
+        """
+        Parameters
+        ----------
+        dfa : NFA
+            `self.nfa`
+        input_str : str
+            `self.input_str`
+        **kwargs
+            The arguments for the `Scene`. Better to keep the default.
+        """
+        super().__init__(**kwargs)
+        self.nfa = nfa
+        self.nfa_graph = _FAGraph(nfa)
+        self.input_str = input_str
+        self.input_symbols = _ManimInput(self.input_str)
+
+    def setup(self) -> None:
+        """Put the diagram and the input string on the screen."""
+        self.add(self.nfa_graph.graph)
+        self.add(self.input_symbols)
+        self.nfa_graph.graph[None].set_color(Animate.HIGHLIGHT_COLOR)
+
+    def construct(self) -> None:
+        """Construct the animation of `self.nfa` identifying `self.input_str`."""
+        self.play(Animate.highlight(self.nfa_graph.graph[None, self.nfa.initial_state]))
+        self.play(
+            Animate.to_default_color(self.nfa_graph.graph[None]),
+            Animate.highlight(self.nfa_graph.graph[self.nfa.initial_state]),
+        )
+        current_transitions: tuple[tuple[Optional[NFAStateT], NFAStateT], ...] = (
+            (None, self.nfa.initial_state),
+        )
+        current_states = {self.nfa.initial_state}
+
+        def add_lambda_transitions() -> Iterable[tuple["NFAStateT", "NFAStateT"]]:
+            """
+            Add states reachable with lambda transitions into `current_states` and
+
+            Returns
+            -------
+            Iterable[tuple[NFAStateT, NFAStateT]]
+                The lambda transitions that can reach new states from one of
+                `current_states`.
+            """
+            states_queue = deque(current_states)
+            while states_queue:
+                state = states_queue.popleft()
+                if next_states := self.nfa.transitions[state].get(""):
+                    for next_state in next_states:
+                        if next_state not in current_states:
+                            current_states.add(next_state)
+                            states_queue.append(next_state)
+                            yield (state, next_state)
+
+        if lambda_transitions := tuple(add_lambda_transitions()):
+            self.play(
+                *self.nfa_graph.change_transitions(
+                    current_transitions, lambda_transitions
+                ),
+                *self.nfa_graph.highlight_states(
+                    map(itemgetter(1), lambda_transitions)
+                ),
+            )
+            current_transitions = lambda_transitions
+        self.wait()
+        input_index = -1
+        for input_index in range(len(self.input_str)):
```

Comment:
```suggestion
        for input_index, input_symbol in enumerate(self.input_str):
```
Avoid the lookup of the input symbol by the index in every iteration. Also, delete the line above (no need to define this variable before the for loop).